protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,294 @@
+        let pending_event = self.poll_pending_events();
+        if pending_event.is_ready() {
+            return pending_event;
+        }
+        if self.ticker.ready() && !self.known_servers.is_empty() {
+            let mut entries = self.address_candidates.drain().collect::<Vec<_>>();
+            entries.sort_unstable_by_key(|(_, count)| *count);
+            let addrs = entries
+                .into_iter()
+                .rev()
+                .map(|(addr, _)| addr)
+                .take(self.config.max_addrs_count)
+                .collect::<Vec<_>>();
+            let peers = if self.known_servers.len() < self.config.test_server_count {
+                self.known_servers.clone()
+            } else {
+                self.known_servers
+                    .choose_multiple(&mut self.rng, self.config.test_server_count)
+                    .copied()
+                    .collect()
+            };
+            for peer in peers {
+                let nonce = self.rng.gen();
+                let req = DialRequest {
+                    nonce,
+                    addrs: addrs.clone(),
+                };
+                self.pending_nonces.insert(nonce);
+                self.submit_req_for_peer(peer, req);
+            }
+            let pending_event = self.poll_pending_events();
+            if pending_event.is_ready() {
+                return pending_event;
+            }
+        }
+        Poll::Pending
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore + 'static,
+{
+    fn submit_req_for_peer(&mut self, peer: PeerId, req: DialRequest) {
+        if let Some(conn_id) = self.peers_to_handlers.get(&peer) {
+            self.pending_events.push_back(ToSwarm::NotifyHandler {
+                peer_id: peer,
+                handler: NotifyHandler::One(*conn_id),
+                event: Either::Left(dial_request::FromBehaviour::PerformRequest(req)),
+            });
+        } else {
+            tracing::debug!(
+                "There should be a connection to {:?}, but there isn't",
+                peer
+            );
+        }
+    }
+
+    fn handle_no_connection(&mut self, peer_id: PeerId, connection_id: ConnectionId) {
+        if matches!(self.peers_to_handlers.get(&peer_id), Some(conn_id) if *conn_id == connection_id)
+        {
+            self.peers_to_handlers.remove(&peer_id);
+        }
+        self.known_servers.retain(|p| p != &peer_id);
+    }
+
+    fn poll_pending_events(
+        &mut self,
+    ) -> Poll<
+        ToSwarm<<Self as NetworkBehaviour>::ToSwarm, <Handler as ConnectionHandler>::FromBehaviour>,
+    > {
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        Poll::Pending
+    }
+}
+
+fn addr_is_local(addr: &Multiaddr) -> bool {
+    addr.iter().any(|c| match c {
+        Protocol::Dns(addr)
+        | Protocol::Dns4(addr)
+        | Protocol::Dns6(addr)
+        | Protocol::Dnsaddr(addr) => addr.ends_with(".local"),
```

Comment:
Pretty sure this can never be hit because the addresses reported to the behaviour in the `handle_` callbacks always come from the interface and thus should always be IPs.

Reply:
Yes you are right, but what if someone would inject an address candidate?

Reply:
Then it will just result in a failed probe right?

Reply:
Are you sure? Why would the probe fail?

You are probably right, so I'll remove it.